cli/internal/fs/path.go

```diff
@@ -50,15 +53,44 @@ func (ap AbsolutePath) Dir() AbsolutePath {
 func (ap AbsolutePath) MkdirAll() error {
 	return os.MkdirAll(ap.asString(), DirPermissions)
 }
-func (ap AbsolutePath) Remove() error {
-	return os.Remove(ap.asString())
-}
 func (ap AbsolutePath) Open() (*os.File, error) {
 	return os.Open(ap.asString())
 }
-func (ap AbsolutePath) ReadFile() ([]byte, error) {
-	return ioutil.ReadFile(ap.asString())
-}
+
 func (ap AbsolutePath) FileExists() bool {
 	return FileExists(ap.asString())
 }
+
+// EnsureDirFS ensures that the directory containing the given filename is created
+func EnsureDirFS(fs afero.Fs, filename AbsolutePath) error {
+	dir := filename.Dir()
+	err := fs.MkdirAll(dir.asString(), DirPermissions)
+	if errors.Is(err, syscall.ENOTDIR) {
+		err = fs.Remove(dir.asString())
+		if err != nil {
+			return errors.Wrapf(err, "removing existing file at %v before creating directories", dir)
+		}
+		err = fs.MkdirAll(dir.asString(), DirPermissions)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return errors.Wrapf(err, "creating directories at %v", dir)
+	}
+	return nil
+}
+
+// WriteFile writes the given bytes to the specified file
+func WriteFile(fs afero.Fs, filename AbsolutePath, toWrite []byte, mode os.FileMode) error {
```

Comment:
Can we switch all of these to be `afero.IOFS`? At that point we can have a common interface of `return fs.MethodName...`

Reply:
Yes, since `IOFS` wraps `Fs`, but `IOFS` is read-only because `io/fs.FS` is read-only. So the functions that do writing won't exactly fit the pattern.

Reply:
I started doing this, but ran into some issues. `afero.ReadFile` has different behavior from `myIOFS.ReadFile()` regarding the leading slash, due to the call to [`ValidPath`](https://pkg.go.dev/io/fs#ValidPath) [here](https://github.com/spf13/afero/blob/master/iofs.go#L85). 

It looks like Go's `io/fs.FS` does not allow absolute paths, and further digging shows that `io/fs.FS` expects non-absolute, platform-independent paths (e.g. `ToSlash()` has been called). So it is not a candidate for a direct pass-through from our _current_ implementation of `AbsolutePath`, which is a platform-dependent absolute path. 

I can think of a few options:
 * Store `fs.AbsolutePath` in a manner consistent with `io/fs.FS`. I think this would be fine w/ 100% coverage of our filesystem operations, which is the goal anyways
 * do conversions when using an instance of `io/fs.FS`. This might be ok, but it makes me nervous.
 * I don't think it's practical to avoid `io/fs.FS`? At least not without changes to `doublestar`? It looks like it's a little bit of an accident that our current usage of `doublestar` isn't triggering this, it's based on all of the instances of `io/fs.FS` that we've passed in so far also implement `io/fs.ReadDir`, which [short-circuits the call](https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/io/fs/readdir.go;l=28) to `Open()`, which would fail.

Reply:
This was the one that I was hoping would work out and ... lol.

Seems like the best ... *path* ... forward here will be to make this sort of change after we've got path types through the entire system. We've got a large number of path types that we need; I'mma get a PR up.